File: frontend/src/app/shared/models/specialization.model.ts
import { CourseModel } from './course.model';

export class SpecializationModel {
  id: number | undefined;
  specialization: string | undefined;
  active: boolean | undefined;
  courseTypes:
    | {
        allow_addition?: boolean;
        course_type?: string;
        credits?: any;
        courses?: any[];
      }[]
    | undefined;
  courses: SpecializationCourseModel[] | undefined;

  constructor(specializationModel: { [key: string]: any } | undefined) {
    this.id = specializationModel?.id;
    this.specialization = specializationModel?.specialization;
    this.courseTypes = specializationModel?.course_types;
    this.courses = specializationModel?.courses?.map(
      (course: { [key: string]: any } | undefined) =>
        new SpecializationCourseModel({ ...course })
    );
    this.active = specializationModel?.active;
  }
}

export class SpecializationCourseModel {
  id: number | undefined;
  course: CourseModel | undefined;
  specialization: SpecializationModel | undefined;
  courseType: string | undefined;

  constructor(specializationCourseModel: { [key: string]: any } | undefined) {
    this.id = specializationCourseModel?.id;
    this.course = new CourseModel({ ...specializationCourseModel?.course });
    this.specialization = new SpecializationModel({
      ...specializationCourseModel?.specialization,
    });
    this.courseType = specializationCourseModel?.course_type;
  }
}

File: backend/models/SpecializationModel.py
import sys
from .db import db
import json

# from models.CourseModel import SpecializationCourse

from sqlalchemy_utils import JSONType

#Creates the Specialization entity with the fields that the database is expecting
class Specialization(db.Model):
    __tablename__ = 'specialization'
    id = db.Column(db.Integer, primary_key=True)
    specialization = db.Column(db.String(150), nullable=False)
    active = db.Column(db.Boolean, default=False, nullable=False)

    # returns the kinds of courses (electives, core) and how many is required
    course_types = db.Column(JSONType)

    courses = db.relationship('SpecializationCourse', backref='specialization')

    def get_course_types(self):
        course_types = json.dumps(self.course_types)
        return course_types

#Creates the SpecializationCourse entity with the fields that the database is expecting
class SpecializationCourse(db.Model):
    __tablename__ = 'specialization_course'
    id = db.Column(db.Integer, primary_key=True)
    course_id = db.Column(db.Integer, db.ForeignKey('course.id'), nullable=False)
    specialization_id = db.Column(db.Integer, db.ForeignKey('specialization.id'), nullable=False)
    course_type = db.Column(db.String(150))
    course = db.relationship('Course', backref='specialization_course')
